# **Exercise 2 Python**

Introduction to Software & Data Engineering with Python - Sekolah Engineering - Pacmann Academy

Materi yang diujikan pada exercise ini adalah
1. Composite Data Structure
2. Looping
3. Function I
4. Function II

# **Composite Data Structure**

### 1

Daftar nilai ujian sebuah kelas adalah sebagai berikut

`nilai = 80, 65, 70, 90, 60, 90, 90, 85, 80, 70, 75, 65, 80, 90, 65, 90` 

##### **1A**

Buatlah sebuah list untuk menampung nilai ujian tersebut. Kemudian hitunglah jumlah murid dan carilah nilai tertinggi dan terendah

In [1]:
# buat list yang menampung nilai ujian

nilai = [80, 65, 70, 90, 60, 90, 90, 85, 80, 70, 75, 65, 80, 90, 65, 90]
print(nilai)

# Hitung jumlah murid di kelas tersebut

jumlah_murid = len(nilai)
print(f"Jumlah murid: {jumlah_murid}")

# print jumlah siswa di dalam kelas

nilai = [80, 65, 70, 90, 60, 90, 90, 85, 80, 70, 75, 65, 80, 90, 65, 90]

jumlah_siswa = len(set(nilai))
print(f"Jumlah siswa dalam kelas: {jumlah_siswa}")

# buat variabel dan print nilai tertinggi

nilai = [80, 65, 70, 90, 60, 90, 90, 85, 80, 70, 75, 65, 80, 90, 65, 90]

nilai_tertinggi = max(nilai)
print(f"Nilai tertinggi dalam kelas: {nilai_tertinggi}")

# buat variabel dan print nilai terendah

nilai = [80, 65, 70, 90, 60, 90, 90, 85, 80, 70, 75, 65, 80, 90, 65, 90]

nilai_tertinggi = min(nilai)
print(f"Nilai tertinggi dalam kelas: {nilai_tertinggi}")

[80, 65, 70, 90, 60, 90, 90, 85, 80, 70, 75, 65, 80, 90, 65, 90]
Jumlah murid: 16
Jumlah siswa dalam kelas: 7
Nilai tertinggi dalam kelas: 90
Nilai tertinggi dalam kelas: 60


##### **1B**

Jika siswa memiliki nilai yang kurang dari rata-rata maka siswa harus ujian ulang. Berapakah jumlah siswa yang memiliki nilai dibawah rata-rata


```
Keterangan:
- Bulatkan nilai rata-rata dua nilai dibelakang koma
- gunakan list comprehension untuk nilai yang kurang dari rata-rata
```



In [126]:
# Nilai ujian

nilai = [80, 65, 70, 90, 60, 90, 90, 85, 80, 70, 75, 65, 80, 90, 65, 90]

# Hitung dan print nilai rata-rata ujian

rata_rata = sum(nilai) / len(nilai)
print(f"Nilai rata-rata ujian: {rata_rata:.2f}")

# Temukan dan print nilai siswa yang kurang dari rata-rata menggunakan list comprehension

nilai_kurang_dari_rata_rata = [nilai_siswa for nilai_siswa in nilai if nilai_siswa < rata_rata]
print("Nilai siswa yang kurang dari rata-rata:", nilai_kurang_dari_rata_rata)

# Hitung dan print jumlah nilai yang kurang dari rata-rata

jumlah_nilai_kurang_dari_rata_rata = len(nilai_kurang_dari_rata_rata)
print(f"Jumlah nilai yang kurang dari rata-rata: {jumlah_nilai_kurang_dari_rata_rata}")


Nilai rata-rata ujian: 77.81
Nilai siswa yang kurang dari rata-rata: [65, 70, 60, 70, 75, 65, 65]
Jumlah nilai yang kurang dari rata-rata: 7


#### **1C**

Hitunglah prosentase siswa yang lulus dan siswa yang tidak lulus ujian.


In [188]:
# Nilai ujian
nilai = [80, 65, 70, 90, 60, 90, 90, 85, 80, 70, 75, 65, 80, 90, 65, 90]

# Tentukan nilai batas kelulusan (misalnya, 70)

batas_kelulusan = 70

# Hitung jumlah siswa yang lulus

siswa_lulus = sum(1 for nilai_siswa in nilai if nilai_siswa >= batas_kelulusan)

# Hitung prosentase siswa yang lulus

prosentase_lulus = (siswa_lulus / len(nilai)) * 100

# Print prosentase siswa yang lulus, bulatkan dua angka dibelakang koma

print(f"Prosentase siswa yang lulus ujian: {prosentase_lulus:.2f}%")

# Hitung prosentase siswa yang tidak lulus

prosentase_tidak_lulus = 100 - prosentase_lulus

# Print prosentase siswa yang tidak lulus

print(f"Prosentase siswa yang tidak lulus ujian: {prosentase_tidak_lulus:.2f}%")


Prosentase siswa yang lulus ujian: 75.00%
Prosentase siswa yang tidak lulus ujian: 25.00%


---

### **2**

Penjualan suatu produk pestisida selama tiga bulan dari tiga jenis produk disimpan di dalam data berikut.

|Bulan|Produk 1|Produk 2|Produk 3|
|-|-|-|-|
|Bulan 1|40|45|39|
|Bulan 2|48|50|43|
|Bulan 3|45|48|45|

Angka pada data di atas menunjukkan jumlah produk yang terjual

#### **2A**

Buatlah sebuah tuple of tuple yang menyimpan seluruh data penjualan produk tersebut per bulan. Simpan hanya angkanya saja.


```
Keterangan:
- Struktur tuple
  - my_tuple = ((jumlah_produk_bulan_1, jumlah_produk_bulan_2, jumlah_produk_bulan_3), ...)
```



In [249]:
# definisikan sebuah tuple

penjualan_pestisida = ((40,45,39),(48,50,43),(45,48,45))

# tampilkan tuple tersebut
print(penjualan_pestisida)

((40, 45, 39), (48, 50, 43), (45, 48, 45))


#### **2B**

Carilah rata-rata penjualan setiap bulan dari ketiga produk tersebut.



```
Keterangan:
- bulatkan hasil penjualan menjadi bilangan bulat
```



In [250]:
penjualan_pestisida = ((40, 45, 39), (48, 50, 43), (45, 48, 45))

# Variabel dari soal 2A

penjualan_bulan_1 = penjualan_pestisida[0]
penjualan_bulan_2 = penjualan_pestisida[1]
penjualan_bulan_3 = penjualan_pestisida[2]

# Hitung dan tampilkan rata-rata penjualan bulan 1

rata_rata_bulan_1 = sum(penjualan_bulan_1) / len(penjualan_bulan_1)
print(f"Rata-rata penjualan bulan 1: {rata_rata_bulan_1:.2f}")

# Hitung dan tampilkan rata-rata penjualan bulan 2

rata_rata_bulan_2 = sum(penjualan_bulan_2) / len(penjualan_bulan_2)
print(f"Rata-rata penjualan bulan 2: {rata_rata_bulan_2:.2f}")

# Hitung dan tampilkan rata-rata penjualan bulan 3

rata_rata_bulan_3 = sum(penjualan_bulan_3) / len(penjualan_bulan_3)
print(f"Rata-rata penjualan bulan 3: {rata_rata_bulan_3:.2f}")


Ellipsis

#### **2C**

Cari rata-rata penjualan setiap produk selama tiga bulan. Gunakan variabel tuple dari soal 2A.



```
Keterangan:
 - untuk akses tuple of tuple gunakan indeks ganda
 - misalkan ada tuple my_tupe = ((1, 2, 3), (4, 5, 6))
 - misalkan ambil angka 2
    - my_tuple[0][1]
 - indexing pertama untuk mengambil index tuple terluar
 - indexing kedua untuk mengambil nilai yang ada di dalam

```



In [251]:
# Variabel dari soal 2A
penjualan_bulan_1 = penjualan_pestisida[0]
penjualan_bulan_2 = penjualan_pestisida[1]
penjualan_bulan_3 = penjualan_pestisida[2]

# Hitung dan print rata-rata penjualan produk 1 selama tiga bulan
rata_rata_produk_1 = sum(penjualan_bulan_1) / len(penjualan_bulan_1)
print(f"Rata-rata penjualan produk 1: {rata_rata_produk_1:.2f}")

# Hitung dan print rata-rata penjualan produk 2 selama tiga bulan
rata_rata_produk_2 = sum(penjualan_bulan_2) / len(penjualan_bulan_2)
print(f"Rata-rata penjualan produk 2: {rata_rata_produk_2:.2f}")

# Hitung dan print rata-rata penjualan produk 3 selama tiga bulan
rata_rata_produk_3 = sum(penjualan_bulan_3) / len(penjualan_bulan_3)
print(f"Rata-rata penjualan produk 3: {rata_rata_produk_3:.2f}")


Ellipsis

### **3**

Lembaga scouting dari sebuah club sepak bola sedang mencari talent muda untuk club mereka. Ada 5 pemain yang masuk incaran, data pemain tersebut sebagai berikut.

|Nama Pemain|Fisik|Mental|Teknik|Menyerang|Bertahan|Posisi|
|-|-|-|-|-|-|-|
|Erling Haaland|89|90|75|95|60|Striker
|Christopher Nkunku|76|78|80|82|65|Sayap Kiri
|Rafael Leao|86|83|85|90|55|Sayap Kiri
|Kevin De Bruyne|82|80|89|88|68|Gelandang Serang
|Vinicius Junior|85|80|85|90|55|Sayap Kiri




#### **3A**

Buatlah sebuah dictionary untuk menyimpan data pemain berikut


In [252]:
# definisikan sebuah dictionary

erling_haaland = {
    'fisik' : 89,
    'mental' : 90,
    'teknik' : 75,
    'menyerang' : 95,
    'bertahan' : 60,
    'posisi' : 'striker'
}
christopher_nkunku = {
    'fisik' : 76,
    'mental' : 78,
    'teknik' : 80,
    'menyerang' : 82,
    'bertahan' : 65,
    'posisi' : 'sayap kiri'
}
rafael_leao = {
    'fisik' : 86,
    'mental' : 83,
    'teknik' : 85,
    'menyerang' : 90,
    'bertahan' : 55,
    'posisi' : 'sayap kiri'
}
kevin_de_bruyne = {
    'fisik' : 82,
    'mental' : 80,
    'teknik' : 89,
    'menyerang' : 88,
    'bertahan' : 68,
    'posisi' : 'gelandang serang'
}
vinicius_junior = {
    'fisik' : 85,
    'mental' : 80,
    'teknik' : 85,
    'menyerang' : 90,
    'bertahan' : 55,
    'posisi' : 'sayap kiri'
}

# print dictionary tersebut
print(erling_haaland)
print(christopher_nkunku)
print(rafael_leao)
print(kevin_de_bruyne)
print(vinicius_junior)

#### **3B**

Hitunglah rata-rata kemampuan masing-masing pemain.


```
Keterangan
 - Untuk akses kemampuan pemain gunakan nama_dict[Nama_pemain][kemampuan]
```



In [1]:
# Fungsi untuk menghitung rata-rata kemampuan
def hitung_rata_rata(kemampuan):
    return sum(kemampuan.values()) / len(kemampuan)

# Pemain-pemain
erling_haaland = {
    'fisik': 89,
    'mental': 90,
    'teknik': 75,
    'menyerang': 95,
    'bertahan': 60,
    'posisi': 'striker'
}

christopher_nkunku = {
    'fisik': 76,
    'mental': 78,
    'teknik': 80,
    'menyerang': 82,
    'bertahan': 65,
    'posisi': 'sayap kiri'
}

rafael_leao = {
    'fisik': 86,
    'mental': 83,
    'teknik': 85,
    'menyerang': 90,
    'bertahan': 55,
    'posisi': 'sayap kiri'
}

kevin_de_bruyne = {
    'fisik': 82,
    'mental': 80,
    'teknik': 89,
    'menyerang': 88,
    'bertahan': 68,
    'posisi': 'gelandang serang'
}

vinicius_junior = {
    'fisik': 85,
    'mental': 80,
    'teknik': 85,
    'menyerang': 90,
    'bertahan': 55,
    'posisi': 'sayap kiri'
}

# Cetak rata-rata kemampuan untuk setiap pemain
print("Rata-rata kemampuan Erling Haaland:", hitung_rata_rata(erling_haaland))
print("Rata-rata kemampuan Christopher Nkunku:", hitung_rata_rata(christopher_nkunku))
print("Rata-rata kemampuan Rafael Leao:", hitung_rata_rata(rafael_leao))
print("Rata-rata kemampuan Kevin De Bruyne:", hitung_rata_rata(kevin_de_bruyne))
print("Rata-rata kemampuan Vinicius Junior:", hitung_rata_rata(vinicius_junior))


TypeError: unsupported operand type(s) for +: 'int' and 'str'

#### **3C**

Jika dilihat dari rata-rata kemampuan pemain, Christopher Nkunku dan Vinicius Junior merupakan dua yang terendah. Hapus Nkunku dan Vinicius dari dictionary pemain.


```
Keterangan:
 - gunakan method pop()
```



In [2]:
# Pemain-pemain
pemain = {
    'erling_haaland': {
        'fisik': 89,
        'mental': 90,
        'teknik': 75,
        'menyerang': 95,
        'bertahan': 60,
        'posisi': 'striker'
    },
    'christopher_nkunku': {
        'fisik': 76,
        'mental': 78,
        'teknik': 80,
        'menyerang': 82,
        'bertahan': 65,
        'posisi': 'sayap kiri'
    },
    'rafael_leao': {
        'fisik': 86,
        'mental': 83,
        'teknik': 85,
        'menyerang': 90,
        'bertahan': 55,
        'posisi': 'sayap kiri'
    },
    'kevin_de_bruyne': {
        'fisik': 82,
        'mental': 80,
        'teknik': 89,
        'menyerang': 88,
        'bertahan': 68,
        'posisi': 'gelandang serang'
    },
    'vinicius_junior': {
        'fisik': 85,
        'mental': 80,
        'teknik': 85,
        'menyerang': 90,
        'bertahan': 55,
        'posisi': 'sayap kiri'
    }
}

# Hapus pemain menggunakan metode pop
pemain.pop('christopher_nkunku', None)
pemain.pop('vinicius_junior', None)

# Tampilkan dictionary pemain yang baru
print("Dictionary Pemain yang Baru:")
for nama_pemain, data_pemain in pemain.items():
    print(nama_pemain + ":", data_pemain)


Dictionary Pemain yang Baru:
erling_haaland: {'fisik': 89, 'mental': 90, 'teknik': 75, 'menyerang': 95, 'bertahan': 60, 'posisi': 'striker'}
rafael_leao: {'fisik': 86, 'mental': 83, 'teknik': 85, 'menyerang': 90, 'bertahan': 55, 'posisi': 'sayap kiri'}
kevin_de_bruyne: {'fisik': 82, 'mental': 80, 'teknik': 89, 'menyerang': 88, 'bertahan': 68, 'posisi': 'gelandang serang'}


#### **3D**

Tersisa tiga pemain yaitu, Erling Haaland, Rafael Leao dan Kevin De Bruyne.
Jika tim scouting sepak bola tersebut sedang mencari penyerang dengan kemampuan **menyerang** terbaik maka siapakah yang akan di ambil ?


```
Keterangan: 
- struktur list
  - [[kemampuan_menyerang, nama_pemain], ...]
- gunakan method max()
```



In [57]:
# Variabel dari soal 3A
kemampuan_menyerang = [
    erling_haaland['menyerang'],
    christopher_nkunku['menyerang'],
    rafael_leao['menyerang'],
    kevin_de_bruyne['menyerang'],
    vinicius_junior['menyerang']
]

# Cari pemain dengan kemampuan menyerang terbaik
pemain_terbaik = max(zip(kemampuan_menyerang, pemain.keys()))

# Tampilkan pemain dengan kemampuan menyerang terbaik
print(f"Pemain dengan kemampuan menyerang terbaik: {pemain_terbaik[1]} (Skor: {pemain_terbaik[0]})")


Pemain dengan kemampuan menyerang terbaik: erling_haaland (Skor: 95)


###**4**

Pak Banu dan Pak Rozi akan bekerja sama dalam berjualan toko buah. Sebelumnya mereka juga sudah mempunyai toko buah, namun karena alasan tertentu mereka memutuskan untuk bekerja sama. Data buah yag di jual Pak Banu dan Pak Rozi adalah sebagai berikut

```
Pak Banu = Jeruk, Apel, Semangka, Rambutan, Pepaya, Kelengkeng, Alpukat, Sirsat
Pak Rozi = Pisang, Melon, Salak, Mangga, Jeruk, Pepaya, Kelengkeng, Srikaya
```





#### **4A**

Buatlah sebuah set untuk menampung buah yang di jual oleh masing-masing penjual

In [58]:
# Set untuk toko buah Pak Banu
buah_pak_banu = {'Jeruk', 'Apel', 'Semangka', 'Rambutan', 'Pepaya', 'Kelengkeng', 'Alpukat', 'Sirsat'}

# Set untuk toko buah Pak Rozi
buah_pak_rozi = {'Pisang', 'Melon', 'Salak', 'Mangga', 'Jeruk', 'Pepaya', 'Kelengkeng', 'Srikaya'}

# Print set masing-masing toko buah
print("Set untuk Toko Buah Pak Banu:", buah_pak_banu)
print("Set untuk Toko Buah Pak Rozi:", buah_pak_rozi)


#### **4B**

Gabungkan buah yang dijual Pak Banu dan Pak Rozi


```
Keterangan:
- gunakan method union()
```



In [111]:
# Set untuk toko buah Pak Banu
buah_pak_banu = {'Jeruk', 'Apel', 'Semangka', 'Rambutan', 'Pepaya', 'Kelengkeng', 'Alpukat', 'Sirsat'}

# Set untuk toko buah Pak Rozi
buah_pak_rozi = {'Pisang', 'Melon', 'Salak', 'Mangga', 'Jeruk', 'Pepaya', 'Kelengkeng', 'Srikaya'}

# Gabungkan set buah yang dijual oleh Pak Banu dan Pak Rozi
buah_gabungan = buah_pak_banu.union(buah_pak_rozi)

# Tampilkan buah yang dijual (gabungan)
print("Buah yang dijual (Gabungan):", buah_gabungan)


Buah yang dijual (Gabungan): {'Sirsat', 'Pisang', 'Semangka', 'Srikaya', 'Alpukat', 'Apel', 'Pepaya', 'Kelengkeng', 'Rambutan', 'Salak', 'Jeruk', 'Melon', 'Mangga'}


#### **4C**

Untuk mempermudah stok ulang buah carilah buah yang sebelumnya sama-sama dijual oleh Pak Banu dan Pak Rozi

```
Keterangan:
- gunakan method intersection()
```

In [112]:
# Set untuk toko buah Pak Banu
buah_pak_banu = {'Jeruk', 'Apel', 'Semangka', 'Rambutan', 'Pepaya', 'Kelengkeng', 'Alpukat', 'Sirsat'}

# Set untuk toko buah Pak Rozi
buah_pak_rozi = {'Pisang', 'Melon', 'Salak', 'Mangga', 'Jeruk', 'Pepaya', 'Kelengkeng', 'Srikaya'}

# Temukan buah yang sama antara Pak Banu dan Pak Rozi
buah_sama = buah_pak_banu.intersection(buah_pak_rozi)

# Tampilkan buah yang sama antara Pak Banu dan Pak Rozi
print("Buah yang dijual oleh Pak Banu:", buah_pak_banu)
print("Buah yang dijual oleh Pak Rozi:", buah_pak_rozi)
print("Buah yang sama antara Pak Banu dan Pak Rozi:", buah_sama)


# **Looping**

### **1**


Pada sebuah puskesmas sedang mengadakan vaksinasi dan memberikan nomor antrian sepanjang 200 orang kepada masyarakat yang ingin vaksin. Untuk **mempercepat** proses vaksinasi puskesmas tersebut **membagi** menjadi empat antrian.

#### **1A**

Untuk nomor antrian kurang dari 100 dibagi berdasarkan bilangan genap dan ganjil. Buatlah urutan bilangan genap dan ganjil menggunakan for loop.

In [163]:
# Inisialisasi list untuk nomor antrian
nomor_antrian = list(range(1, 201))

# Inisialisasi list untuk nomor antrian genap dan ganjil
antrian_genap = []
antrian_ganjil = []

# Pembagian antrian berdasarkan bilangan genap dan ganjil
for nomor in nomor_antrian:
    if nomor % 2 == 0:
        antrian_genap.append(nomor)
    else:
        antrian_ganjil.append(nomor)

# Print urutan bilangan genap
print("Urutan Bilangan Genap:")
for genap in antrian_genap:
    print(genap, end=' ')

# Print urutan bilangan ganjil
print("\nUrutan Bilangan Ganjil:")
for ganjil in antrian_ganjil:
    print(ganjil, end=' ')


Urutan Bilangan Genap:
2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50 52 54 56 58 60 62 64 66 68 70 72 74 76 78 80 82 84 86 88 90 92 94 96 98 100 102 104 106 108 110 112 114 116 118 120 122 124 126 128 130 132 134 136 138 140 142 144 146 148 150 152 154 156 158 160 162 164 166 168 170 172 174 176 178 180 182 184 186 188 190 192 194 196 198 200 
Urutan Bilangan Ganjil:
1 3 5 7 9 11 13 15 17 19 21 23 25 27 29 31 33 35 37 39 41 43 45 47 49 51 53 55 57 59 61 63 65 67 69 71 73 75 77 79 81 83 85 87 89 91 93 95 97 99 101 103 105 107 109 111 113 115 117 119 121 123 125 127 129 131 133 135 137 139 141 143 145 147 149 151 153 155 157 159 161 163 165 167 169 171 173 175 177 179 181 183 185 187 189 191 193 195 197 199 

#### **1B**

Untuk nomor urut 101 sampai 200 dibagi berdasarkan nomor urut
 
 - `Nomor 101 - 125 dan 151 - 175`
 - `Nomor 126 - 150 dan 175 - 200`

Buatlah deret yang menampung dua kondisi tersebut

In [164]:
# Inisialisasi list untuk nomor antrian 101-200
nomor_antrian_101_200 = list(range(101, 201))

# Inisialisasi list untuk dua kondisi
kondisi_1 = []
kondisi_2 = []

# Pembagian nomor antrian berdasarkan dua kondisi
for nomor in nomor_antrian_101_200:
    if (nomor >= 101 and nomor <= 125) or (nomor >= 151 and nomor <= 175):
        kondisi_1.append(nomor)
    elif (nomor >= 126 and nomor <= 150) or (nomor >= 176 and nomor <= 200):
        kondisi_2.append(nomor)

# Print kedua deret tersebut
print("Deret untuk Kondisi 1:")
for bilangan_kondisi_1 in kondisi_1:
    print(bilangan_kondisi_1, end=' ')

print("\nDeret untuk Kondisi 2:")
for bilangan_kondisi_2 in kondisi_2:
    print(bilangan_kondisi_2, end=' ')


### **2**

Armada Angkutan Bus Setya Jaya mengeluh karena ada kenaikan harga BBM. Tarif angkutan sebelum kenaikan BBM bisa dilihat pada tabel biaya perjalanan dibawah

|Jurusan | Biaya per Orang | Biaya Pelajar|
|-|-|-|
|Terminal A ke Terminal B | 5.000| 2.500|
|Terminal A ke Terminal C | 10.000 | 4.000|
|Terminal A ke Terminal D | 15.000 | 7.000|


Untuk mengimbangi pengeluaran dan pendapatan para pengusaha angkutan sepakat untuk menaikkan harga.

#### **2A**

Buatlah list yang menampung biaya per orang dan biaya pelajar

In [165]:
# definisikan dua list yang menampung biaya per orang dan biaya pelajar
...

# print kedua list tersebut
...

#### **2B**

Jika **biaya per orang** naik 5.000 rupiah pada semua jurusan terminal, berapa tarif per orang sekarang. Hitung dengan menggunakan **while loop**

In [166]:
# gunakan list yang sudah didefinisikan dari soal 2A

# buat variabel untuk menghentikan while
...

# hitung tarif baru dengan while
...

# tampilkan list tarif penumpang yang baru
...

Ellipsis

#### **2C**

Jika tarif penumpang **Pelajar** naik 2.000 rupiah pada semua jurusan terminal, berapa tarif pelajar sekarang ? Hitung menggunakan while loop

In [167]:
# gunakan list yang sudah didefinisikan dari soal 2A

# buat variabel untuk menghentikan while
...

# hitung tarif baru dengan while
...

# tampilkan list tarif penumpang pelajar yang baru
...

Ellipsis

### **3**

Lembaga scouting dari sebuah club sepak bola sedang mencari talent muda berbakat untuk club mereka. Ada 5 pemain yang masuk incaran, data pemain tersebut sebagai berikut.

|Nama Pemain|Fisik|Mental|Teknik|Menyerang|Bertahan|
|-|-|-|-|-|-|
|Erling Haaland|89|90|75|95|60|
|Christopher Nkunku|76|78|80|82|65|
|Rafael Leao|86|83|85|90|55|
|Kevin De Bruyne|82|80|89|88|68|
|Vinicius Junior|85|80|85|90|55|

#### **3A**

Buatlah dictionary untuk menyimpan data pemain berikut


```
Keterangan:
- Struktur dictionary
  - dict = {nama_pemain:{kemampuan_1:nilai, kemampuan_2:nilai, ...}, ...}
```



In [168]:
# definisikan sebuah dictionary
...
# print dictionary tersebut
...

#### **3B**

Club membutuhkan pemain bola dengan mental yang bagus. Jika ada pemain dengan **mental** lebih kecil dari **80** keluarkan dari data scouting.

In [169]:
# untuk menghindari eror pada saat loop, gunakan method copy untuk menyalin dictionary pada variabel yang baru
# buat nama variabel dahulu kemudian copy dictionary yang sudah dibuat sebelumnya
...

# gunakan for dan if else untuk menyeleksi pemain
# jika pemain masuk memenuhi syarat pakai "continue"
# jika tidak memenuhi syarat, langsung keluarkan dari dict, pakai method pop()
...

# gunakan list comprehension untuk menampilkan pemain yang tersisa
# buat variabel baru dahulu untuk menyimpan hasil list comprehension
...

Ellipsis

#### **3C**


Pengamat Sepak Bola menyarankan untuk langsung mengambil pemain dengan kemampuan **menyerang** lebih dari **92**. Namun pihak scouting masih mempertimbangkan hal lain. Untuk itu pihak scouting hanya akan melihat adakah pemain dengan kemampuan lebih dari **92**.

In [170]:
# gunakan cara yang sama agar saat loop dictionary tidak eror yaitu menyalin dictionary ke variabel baru
# bedakan nama variabel copy dengan yang sebelumnya
...

# pakai for loop dan if else untuk mencari pemain dengan kemampuan menyerang lebih dari 92
# pakai pass jika menemukan pemain dengan kondisi tersebut
...
    

Ellipsis

#### **3D**

Pelatih membutuhkan pemain dengan **teknik** lebih dari **85** dan **fisik** lebih dari sama dengan **84**.
Jika ada pemain yang tidak memenuhi syarat tersebut langsung keluarkan dari data scouting.

In [171]:
# gunakan cara yang sama agar saat loop dictionary tidak eror yaitu menyalin dictionary ke variabel baru
# bedakan nama variabel copy dengan yang lain
...

# gunakan for dan if else untuk menyeleksi pemain
# jika pemain masuk memenuhi syarat pakai "continue"
# jika tidak memenuhi syarat, langsung keluarkan dari dict, pakai method pop()
...


Ellipsis

#### **3E**

Masih tersisa dua pemain yaitu Rafael Leao dan Vinicius Junior. Uang kas klub hanya bisa digunakan untuk membeli satu pemain saja, untuk itu cari pemain dengan rata-rata kemampuan paling tinggi dari kedua pemain tersebut

In [172]:
# gunakan cara yang sama agar saat loop dictionary tidak eror yaitu menyalin dictionary ke variabel baru
# bedakan nama variabel copy dengan yang lain
...

# definisikan variabel untuk menampung rata-rata
...

# gunakan for untuk menyeleksi pemain
# didalam for yang pertama, definisikan variabel jumlah untuk menampung jumlah total kemampuan pemain
# anda bisa memakai method keys atau cara lain pada for yang kedua jika anda pakai nested-for
...

Ellipsis

### **4**

Bu Surti akan berbelanja di swalayan dengan Bu Mima dan Bu Rani


```
belanjaan ibu surti = 20.000, 80.000, 40.000, 150.500
belanjaan ibu mima = 31.000, 43.000, 65.000, 105.000
belanjaan ibu rani = 12.000, 200.000, 61.000, 10.000
```

#### **4A**

Buatlah list untuk menampung harga belanja masing-masing ibu tersebut

In [173]:
# buat tiga list untuk menampung list harga belanjaan
...

#### **4B**

Cari total dari belanjaan masing-masing ibu dengan list comprehension, jika ada harga yang lebih dari 50.000 diskon 5%

In [174]:
# Cari total belanja dengan list comprehension
# HINT, Gunakan method sum()
...

# tampilkan masing-masing total belanjaan
...

Ellipsis

# **Function I**

---
## Case 1 : Waktu Mencuci Tangan
Dianjurkan bagi seseorang untuk mencuci tangan sampai 21 detik secara menyeluruh agar tangan dapat dikatakan bersih dari kuman dan kotoran. 

Buatlah fungsi yang menerima input jumlah seseorang mencuci tangan dalam satu hari dan lama waktu orang tersebut mengikuti anjuran cuci tangan (dalam bulan). 

Hitung banyaknya waktu (dalam menit dan detik) yang dihabiskan orang tersebut untuk mencuci tangan.   

- anggap 1 bulan == 30 hari

```
Contoh:
Setelah mengetahui anjuran tersebut, selama 2 bulan terakhir Athaya mencuci tangan sesuai dengan anjuran sebanyak 6 kali dalam sehari

input  
6 kali dan 2 bulan

ouput :
Waktu total mencuci tangan adalah 126 menit dan 0 detik

```

### 1A.
Buatlah fungsi tersebut!
```
Catatan tambahan:
  - Gunakan paramater yang dapat menerima input jumlah cuci tangan dan jumlah bulan
  - Gunakan return pada fungsi untuk mengembalikan output yang didapat
```

In [175]:
#membuat fungsi
...


###1B.
Panggil fungsi tersebut! Apabila,

a. ) Selama 1 tahun 1 bulan terakhir Richard mencuci tangan sesuai dengan anjuran sebanyak 3 kali dalam sehari

In [176]:
#memanggil fungsi 
...

b. ) Selama 3 bulan terakhir, Nabila mencuci tangan sesuai anjaran sebanyak 5 kali dalam sehari

In [177]:
#memanggil fungsi 
...

---
##Case 2 : Perkalian
Buatlah fungsi yang menghasilkan list hasil perkalian! 

Sebagai contoh jika user meng-inputkan angka 7. Berarti fungsi akan menampilkan perkalian 7 x 1 sampai 7 x 10.

```
Contoh:
input  
7

ouput :
[7, 14, 21, 28, 35, 42, 49, 56, 63, 70]
```


###2A.
Membuat fungsi **tabel perkalian**.
```
Catatan:
 - Parameter yang digunakan adalah angka dan panjang tabel perkalian
 - Karena panjang tabel perkalian hingga 10. Maka buatlah default value = 10
 - Gunakan return untuk mengembalikan nilai
 - Anda dapat menggunakan looping untuk mengembangkan fungsi
```

In [178]:
#membuat fungsi tabel_perkalian
...

###2B.
Memanggil fungsi **tabel_perkalian** dengan:

a.) Tampilkan tabel perkalian 7

In [179]:
#memanggil fungsi
...

b.) Tampilkan tabel perkalian 121

In [180]:
#memanggil fungsi
...

c.) Kemudian, anda ingin menampilkan tabel perkalian 10 tetapi dengan panjang tabel hingga 15

Tampilkan nilai tersebut tanpa merubah fungsi
```
Catatan:
  - Anda dapat menggunakan keyword arguments
```

In [181]:
#memanggil fungsi
...

---
##Case 3 : Perhitungan BMI
Anda bekerja disuatu pekerjaan dibidang kesehatan dan diminta untuk membuat kalkulator BMI (Body Mass index) sederhana yang dapat menentukan apakah orang tersebut mengalami berat badan dibawah normal, berat badan normal, kelebihan berat badan, atau obesitas berdasarkan berat badan dan tinggi badan yang dimasukkan. Fungsi dikembangkan dengan ketentuan sebagai berikut :

###3A.
Buatlah sebuah fungsi **bmi** yang dapat menampung input dari user yang berisi berapa berat badan user (dalam kilogram) dan berapa tinggi badan user (dalam sentimeter)

```
🔽 Ketika fungsi dijalankan, akan menghasilkan 🔽
 Masukkan berat badan (dalam kg) :
 Masukkan tinggi badan (dalam cm) : 
 
```

In [182]:
#buat fungsi
...

Untuk mengecek apakah fungsi yang anda buat dapat berjalan sesuai kriteria atau tidak, anda dapat memanggil fungsi bmi

In [183]:
#memanggil fungsi
...

###3B.
Kembangkanlah fungsi yang sudah dibuat agar dapat menghitung BMI dari input yang sudah diberikan dan menentukan kategori BMI.

| BMI |Kategori|
|-|-|
|**<18,5**  | Berat Badan dibawah normal| 
|**18,5 - 22,9**  | Berat badan normal | 
|**23 - 24,9**  | Kelebihan berat badan | 
|**>= 25**  | Obesitas | 

```
Keterangan :
 - Input berat badan dan tinggi badan dari user yang akan digunakan untuk melakukan perhitungan BMI
 - Rumus perhitungan BMI = berat badan(kg) : ( tinggi badan(meter) x tinggi badan(meter) )
 - Bulatkan hasil BMI hingga 2 angka dibelakang koma.
 - Anda dapat memanfaatkan branching untuk mengembangkan fungsi.
 - Untuk meminimalisir terjadinya error, ketika user input selain angka / number. Maka program akan menampilkan "Input harus berupa angka” 

🔽Contoh output ketika fungsi dijalankan, akan menghasilkan 🔽
Masukkan berat badan (dalam kg) : 60
Masukkan tinggi badan (dalam cm) :  162

Body Mass Index (BMI) : 22.86
Kategori : Berat normal
```

In [184]:
#Mengembangkan fungsi agar dapat menghitung dan menetukan BMI
...

Untuk dapat mengecek apakah fungsi yang anda buat sesuai dengan kriteria diminta atau tidak, anda dapat melakukan pemanggilan fungsi **bmi**

input A :
Berat badan Ara = 60 kg
Tinggi Badan Ara = 162 cm

input B :
Berat badan Ahmad = 62 kg
Tinggi Badan Ahmad = 187,4 cm

In [185]:
#memanggil fungsi dengan input A
...

In [186]:
#memanggil fungsi dengan input B
...

---
##Case 4
Sebuah sekolah sedang melakukan survey mengenai pengeluaran transportasi (berangkat-pulang sekolah) perbulan yang dikeluarkan siswa SD A selama sebulan 

Untuk menunjang survey pengeluaran, anda diminta untuk membuat fungsi dengan menggunakan arbitrary keyword arguments yang dapat menampilkan pasangan nama siswa dan besaran biaya transport (rupiah) yang dikeluarkan perbulannya.


###4A. 
Membuat fungsi menggunakan arbitrary keyword arguments
```
Catatan:
 - Nilai yang dimasukkan adalah nama siswa dan biaya pengeluaran transportasi perbulannya

🔽 Ketika fungsi dipanggil, output yang diharapkan 🔽
  Pengeluaran Niriina Amalina untuk transportasi senilai : Rp.150000
```

In [187]:
#membuat fungsi
...

###4B.
Beberapa hari kemudian, telah didapat sampel sementara dari survey yang dijabarkan dalam tabel berikut:

| Nama Siswa |Pengeluaran Transportasi Perbulan|
|-|-|
|Randy | Rp. 240.000 | 
|Alfin| Rp. 200.000 | 
|Shafira | Rp. 175.000 | 
|Andi  | Rp. 320.000| 
|Nuning  | Rp. 400.000 | 

Cobalah panggil fungsi berdasarkan data diatas!

In [188]:
#memanggil fungsi
...

---
##Case 5 : Docstring Perhitungan Faktorial
Terdapat fungsi **print_faktor** yang dapat mengetahui faktorisasi dari suatu angka:

In [189]:
def print_faktor(angka):
    print("faktor dari ",angka,"adalah:")
    for i in range(1, angka + 1):
        if angka % i == 0:
            print(i, end = ' ')

angka_input = 50

In [190]:
print_faktor(angka_input)

faktor dari  50 adalah:
1 2 5 10 25 50 

Akan tetapi, fungsi tersebut masih sulit untuk dipahami. Maka dari itu, buatlah docstring dari fungsi diatas agar mempermudah user dalam memahami program!

In [191]:
#membuat docstring
#copy paste fungsi diatas dan tambahkan docstring
...


Ellipsis

##Case 6 : Perhitungan Pendapatan Perkapita
Anda ingin mengetahui tingkat kemakmuran negara dan masyarakat melalui perhitungan Pendapatan PerKapita (PPK). 

Maka anda berencana untuk membuat fungsi **ppk** yang dapat menghitung hasil pendapatan rata-rata masyarakat pada suatu negara

```
Catatan:
 Rumus pendapatan perkapita = pendapatan nasional : jumlah penduduk
```

###6A.
Buatlah fungsi tersebut!

```
Catatan :     
 - Didalam fungsi, argumen yang akan diinputkan adalah nama negara, pendapatan nasional suatu negara, dan total jumlah penduduk.

🔽Contoh output ketika fungsi dijalankan, akan menghasilkan 🔽
Pendapatan negara x adalah y milyar rupiah perpenduduk.
 ```

In [192]:
#membuat fungsi ppk
...

Panggil fungsi tersebut dengan menggunakan data dibawah ini

| input ke -1 | Nama Negara |Pendapatan Nasional (Miliar Rupiah) | Jumlah Penduduk (Juta Jiwa)
| - |-|-| -|
|1. |Indonesia | 210 | 200 |  
|2. |Malaysia | 75 | 25 |
|3. |Vietnam | 80 | 67 | 

In [193]:
# memanggil fungsi dengan input ke-1
...

In [194]:
# memanggil fungsi dengan input ke-2
...

In [195]:
# memanggil fungsi dengan input ke-3
...

###6B.
Kemudian, untuk memudahkan dalam memahami isi dari fungsi, anda berencana untuk membuat docstring fungsi ppk

In [196]:
#fungsi ppk
#membuat docstring dari fungsi ppk
...

Ellipsis

###6C.
Anda berencana untuk membuat fungsi baru ppk dengan menggunakan fungsi lambda agar fungsi lebih ringkas dan sederhana.
```
🔽Contoh output ketika fungsi dijalankan, akan menghasilkan 🔽
Pendapatan negara x adalah y milyar rupiah perpenduduk.
 ```

In [197]:
#membuat fungsi ppk dengan lambda
...

###6D.
Panggil fungsi lambda dari data Indonesia!

In [198]:
# memanggil fungsi 
...

# **Function II**

---
##Case 1 : Kebutuhan Kalori Harian
Sebuah lembaga kesehatan berencana untuk melakukan penelitian mengenai pengaruh kebutuhan kalori / Basal Metabolic Rate (BMR) pada pekerja kantor yang aktivitas sehari-harinya duduk seharian penuh didepan komputer.

Untuk pengembangan proyek penelitian, anda diminta untuk membuat modul python berisi fungsi-fungsi menghitung kebutuhan kalori harian pria dan wanita.

###1A.
Untuk langkah awal, buatlah modul **bmr** yang berisi fungsi perhitungan kebutuhan kalori untuk pria dan wanita.

Kemudian, buatlah fungsi baru **pria** yang berisi perhitungan kebutuhan kalori berdasarkan rumus Harris Beneddict.
```
Catatan :
 - Argumen yang akan diinputkan adalah berat badan (kg), tinggi badan (cm), dan usia(tahun)
 - Rumus menghitung BMR pria = ( 88,362 + (13,397 x berat badan) + (4,799 x tinggi badan) – (5,677 x usia) ) x 1,2
 - Bulatkan hasil BMR dua angka dibelakang koma

🔽 Output ketika program dijalankan, akan menghasilkan 🔽
Kebutuhan kalori yang dibutuhkan sebesar :


```

In [199]:
#modul bmr
#fungsi pria
...

Ellipsis

###1B.
Selanjutnya, buatlah fungsi baru **wanita** yang berisi perhitungan kebutuhan kalori berdasarkan rumus Harris Beneddict.
```
Catatan :
 - Argumen yang akan diinputkan adalah berat badan (kg), tinggi badan (cm), dan usia(tahun)
 - Rumus menghitung BMR wanita = ( 447,593 + (9,247 x berat badan) + (3,098 x tinggi badan) – (4,330 x usia) ) x 1,2
 - Bulatkan hasil BMR dua angka dibelakang koma

🔽 Output ketika program dijalankan, akan menghasilkan 🔽
Kebutuhan kalori yang dibutuhkan sebesar :


```

In [200]:
#modul bmr
# fungsi pria
...

# fungsi wanita
...

Ellipsis

###1C.
Setelah modul disimpan, anda ingin mengecek apakah modul yang anda bangun dapat berjalan atau tidak.

Untuk itu anda berencana mencoba membuat file utama untuk mengecek apakah modul dapat berjalan dengan baik. Impor modul **bmr** yang telah dibuat sebelumnya.



In [201]:
#impor modul
...

###1D.
Coba panggil fungsi **pria** dengan nilai input:
  - Berat badan Wahyu = 70 kg
  - Tinggi Badan Wahyu = 181 kg
  - Usia Wahyu = 25 tahun



In [202]:
#akses fungsi pria
...

###1E.
Coba panggil fungsi **wanita** dengan nilai input:
  - Berat badan Rita = 62 kg
  - Tinggi Badan Rita = 158 kg
  - Usia Rita  = 27 tahun



In [203]:
#akses fungsi wanita
...  

---
##Case 2 : Laporan Penjualan
Terdapat laporan penjualan perusahaan Z di Bulan April yang disajikan dalam bentuk tabel dibawah.

|Laporan Penjualan Bulan April | Keterangan |
|-|-|
|Penjualan | Rp. 100.000.000|
|Potongan Penjualan |Rp. 15.000.000|
|Retur Pembelian | Rp. 2.000.000|
|Persediaan Barang |Rp. 75.000.000|
|Persediaan Akhir |Rp. 15.000.000|


###2A.
Buatlah sebuat file python baru dengan nama **main.py** dan tambahkan dictionary berisi tabel Laporan Penjualan Bulan April.

In [204]:
#file main.py
#dictionary
...

Ellipsis

In [205]:
#coba cek isi dictionary, apakah sudah sesuai dengan data atau belum
...

###2B.
Selanjutnya, anda diminta untuk membuat modul baru dengan nama **laporan penjualan** yang berisi perhitungan harga pokok penjualan(HPP), penjualan bersih, dan laba kotor dari tabel diatas,

Buatlah **fungsi hpp** dari modul laporan penjualan
```
Catatan :
 - Rumus menghitung harga pokok penjualan = persediaan barang - persediaan akhir 
 - Value diambil langsung dari dictionary yang dibuat sebelumnya di main file.
```

In [206]:
#modul laporan_penjualan
#fungsi hpp
...

Ellipsis

###2C.
Kemudian buatlah **fungsi penjualan bersih** dari modul laporan penjualan
```
Catatan :
 - Rumus menghitung penjualan bersih = penjualan - (potongan penjualan + retur pembelian) 
 - Value diambil langsung dari dictionary yang dibuat sebelumnya di main file.
```

In [207]:
# modul laporan_penjualan
# fungsi hpp
...

# fungsi penjualan_bersih
...

Ellipsis

###2D.
Kemudian buatlah **fungsi laba kotor** dari modul laporan penjualan
```
Catatan :
 - Rumus menghitung laba kotor = penjualan bersih - harga pokok penjualan
 - Value diambil langsung dari dictionary yang dibuat sebelumnya di main file.
```

In [208]:
# modul laporan_penjualan
# fungsi hpp
...

# fungsi penjualan_bersih
...

# fungsi laba kotor
...

Ellipsis

###2E.
Simpanlah modul python tersebut. Kemudian, untuk mencoba apakah modul dapat berjalan dengan baik di **main.py**. Import dan jalankan fungsi-fungsi yang ada didalamnya.

Import modul


In [209]:
# main.py
# import modul
...

# dictionary yang telah dibuat sebelumnya
...

Ellipsis

###2F.
Panggil fungsi **hpp** dengan nilai input dari dictionary yang telah dibuat sebelumnya.



In [210]:
# main.py
# import modul
...

# dictionary yang telah dibuat sebelumnya
...
#memanggil fungsi hpp
...

Ellipsis

###2G.
Setelah nilai harga pokok penjualan telah didapat. Simpan nilai hpp kedalam dictionary agar nantinya dapat digunakan kembali oleh program.




In [211]:
# main.py
# import modul
...

# dictionary yang telah dibuat sebelumnya
...

#memanggil fungsi hpp
...

#simpan nilai hpp kedalam dictionary
...

Ellipsis

###2H.
Panggil fungsi **penjualan_bersih** dengan nilai input dari dictionary yang telah dibuat sebelumnya.



In [212]:
#memanggil fungsi penjualan bersih
...